### PR TITLE
Add pre-commit config and actions file

### PR DESCRIPTION
**Pull request by jpthiele**
This adds a pre-commit config with one commit per check/hook.
For end-of-file-fixer and runic the needed changes to make the codebase pass the check
have been applied in the respective commit.

Finally this also adds a new CI workflow file to run pre-commit on all files.

### CODE REVIEW

**Comment by jpthiele** [comment]
What is still needed is a description of how to install and use pre-commit. 
Where do you think this fits better, the Readme or the documentation?

**Comment by pjaap** [comment]
Wow, that's a lot to review :laughing: 

image
(this is usually a programming joke, right?)

However, I scrolled though the changes and Runic does a damn good job, in my opinion.

We should, as you said, clarify how to use the other hooks if someone wants to use them.

I suggest to put this in the README, or even in the ci-config as a comment.

File: src/makie.jl [hunk 1]
```diff
@@ -139,12 +145,12 @@ function scene_interaction(update_scene,
...
     # Set keyboard event callback
-    XMakie.on(scene.events.keyboardbutton) do buttons
+    return XMakie.on(scene.events.keyboardbutton) do buttons
         if _inscene(scene, mouseposition[])
             # On pressing a switch key, pass control
-            for i = 1:length(switchkeys)
+            for i in 1:length(switchkeys)
                 if switchkeys[i] != :nothing &&
-                   XMakie.ispressed(scene, getproperty(XMakie.Keyboard, switchkeys[i]))
+                        XMakie.ispressed(scene, getproperty(XMakie.Keyboard, switchkeys[i]))
```
[the remark this answers is not in the document]
Comment (pjaap) [reply src/makie.jl:153]: Ah, I see.  That line is part of the if expression. Ok, then it makes sense.

**Comment by pjaap** [comment]
I would also say to leave in all the additional checks in the CI. For some of them the meaning is not completely clear to me, but I guess it will when they are triggered in the future. 

I want to add the the output of the checks is pretty nice
image
So we can see which part failes and act apparently. Or remove the failing check.

**Comment by jpthiele** [comment]
> I want to add the the output of the checks is pretty nice 
once enable locally you also get this nice output on calling `git commit`.

**Comment by jpthiele** [comment]
If any of you want to try locally you have to install `pre-commit` either through `conda` or `pip` whichever you prefer. 

Then you have to run `pre-commit install` *once* per repo to opt-in. 

You also need to install Runic.jl in a shared project `@runic` as described on the repo.

Finally, you can run `pre-commit run --all-files` to check beforehand.
It will also run on the set of changed files on every `git commit` before actually making a commit.

**Comment by pjaap** [comment]
`pre-commit` should also be available as a package on most Linux distributions, btw.

**Comment by jpthiele** [comment]
> Will it still be possible to work without installing pre-commit locally - I don't mean myself, but I want to keep the hurdles to contributing low.
> 
Yes, `pre-commit` is completely opt-in!
Well apart from the CI check of course.

> Also hints on that should IMHO should got into some "hints for contirbutors" chapter in the docs or an extra CONTRIBUNTING.md. Or we do it like in LinearSolve and add a "Contributor's" guide badge to README. Alternatively, is there already a "runic" badge ? This could point to some info collected in the runic repo, including the precommit stuff etc.

I agree, but I was not sure where to put it best

**Pull request merged**